internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -757,3 +786,133 @@ func EnqueueForCompositionRevisionFunc(of resource.CompositeKind, list func(ctx
 		}
 	}
 }
+
+// handleFunctionResults will process the result returned by the composition functions.
+// This includes recording events for the xr and claim and setting updating status conditions
+// on the xr.
+func handleFunctionResults(r *Reconciler, log logging.Logger, cm *claim.Unstructured, xr *composite.Unstructured, res CompositionResult) {
+	recordEvents(r, log, cm, res.ClaimEvents)
+	recordEvents(r, log, xr, res.CompositeEvents)
+
+	prevXrConditions := GetConditions(&xr.Unstructured)
+	filteredXrConditions := handleConditions(r.record, xr, prevXrConditions, res.CompositeConditions)
+	ForceSetConditions(&xr.Unstructured, filteredXrConditions...)
+
+	prevCmConditions := GetClaimConditions(xr)
+	filteredCmConditions := handleConditions(r.record, cm, prevCmConditions, res.ClaimConditions)
+	ForceSetClaimConditions(xr, filteredCmConditions...)
+}
+
+func recordEvents(r *Reconciler, log logging.Logger, o runtime.Object, events []event.Event) {
+	if o == nil {
+		// in the case that the object is nil (e.g., failed to fetch claim), do not emit events on this run
+		return
+	}
+	for _, e := range events {
+		log.Debug(e.Message)
+		if o != nil {
+			r.record.Event(o, e)
+		}
+	}
+}
+
+// ForceSetConditions sets the object's status.conditions to whatever is provided
+// as the conditions arg. All existing conditions will be wiped.
+func ForceSetConditions(u *kunstructured.Unstructured, conditions ...xpv1.Condition) {
```

Comment:
Info: Added a lot of functions here to handle condition manipulation for this feature. It seems all similar code exists in the runtime, but not sure if we want to add these to the runtime as they are pretty specific to this one use-case and may encourage runtime users to not follow best practices.